src/option.c

```diff
@@ -7418,34 +7451,55 @@ set_context_in_set_cmd(
 	}
     }
 
-    // For an option that is a list of file names, find the start of the
-    // last file name.
-    for (p = arg + STRLEN(arg) - 1; p > xp->xp_pattern; --p)
+    // For an option that is a list of file names, or comma/colon-separated
+    // values, split it by the delimiter and find the start of the current
+    // pattern, while accounting for backslash-escaped space/commas/colons.
+    // Triple-backslashed escaped file names (e.g. 'path') can also be
+    // delimited by space.
+    if ((flags & P_EXPAND) || (flags & P_COMMA) || (flags & P_COLON))
     {
-	// count number of backslashes before ' ' or ','
-	if (*p == ' ' || *p == ',')
+	for (p = arg + STRLEN(arg) - 1; p >= xp->xp_pattern; --p)
 	{
-	    s = p;
-	    while (s > xp->xp_pattern && *(s - 1) == '\\')
-		--s;
-	    if ((*p == ' ' && (xp->xp_backslash == XP_BS_THREE && (p - s) < 3))
-		    || (*p == ',' && (flags & P_COMMA) && ((p - s) & 1) == 0))
+	    // count number of backslashes before ' ' or ',' or ':'
+	    if (*p == ' ' || *p == ',' ||
+		    (*p == ':' && (flags & P_COLON)))
 	    {
-		xp->xp_pattern = p + 1;
-		break;
+		s = p;
+		while (s > xp->xp_pattern && *(s - 1) == '\\')
+		    --s;
+		if ((*p == ' ' && (xp->xp_backslash == XP_BS_THREE && (p - s) < 3))
+			|| (*p == ',' && (flags & P_COMMA) && ((p - s) % 1) == 0)
```

Comment:
This `(p - s) % 1` looks strange. Isn't any number module 1 always 0?

Reply:
You are absolutely right. I only moved this code around and added the `P_COLON` check below it but I didn't change the logic for handling commas here. In fact, the logic here is wrong. I didn't want the scope of PR to grow any more which is why I didn't fix it.

I think the correct logic here is actually `(p - s) >= 2`. Currently, if you have a file/folder of the name `foo,bar`, the correct way to specify it in an option is do so something like `:set backupdir=dir1,foo\\,bar,dir2`. Note that you will see some tests in my new code that checks that `set-=` will actually that the escaping we do when completing existing option is correctly doing that.

Note: there's another cmd auto-complete bug (that's why I wanted to fix this in another PR instead of growing in scope). In the above example, if you do auto-complete as follows: `set backupdir=fo<Tab>`, if will complete to `set backupdir=foo,bar`, forgetting to insert the `\\` there. In general the option setting syntax is a little weird regarding commas, as I believe the syntax is incomplete and doesn't allow certain types of names to be entered (you cannot enter a folder/file that ends with "\", because that slash will end up escaping the comma instead).

tldr: Yes it's broken, but it's broken before my change and I could submit another PR to fix it.

Reply:
The commas will be fixed in #13303